jormungandr/src/explorer/mod.rs

```diff
@@ -66,112 +89,303 @@ impl Explorer {
 
         let mut explorer_db = self.db.clone();
         let logger = info.logger().clone();
-        let blockchain = self.blockchain.clone();
         match bquery {
-            ExplorerMsg::NewBlock(new_block_ref) => info.spawn(lazy(move || {
-                explorer_db
-                    .store_ref(new_block_ref.clone())
-                    .map_err(|_| unreachable!())
-                    .join(
-                        explorer_db
-                            .index_transactions(new_block_ref, blockchain)
-                            .map_err(move |err| {
-                                error!(logger, "Explorer error: {}", err);
-                            }),
-                    )
-                    .map(move |_| ())
-            })),
+            ExplorerMsg::NewBlock(block) => info.spawn(explorer_db.apply_block(block).then(
+                move |result| match result {
+                    Ok(_gc_root) => Ok(()),
+                    Err(err) => Err(error!(logger, "Explorer error: {}", err)),
+                },
+            )),
         }
         future::ok::<(), ()>(())
     }
 }
 
-#[derive(Clone)]
-pub struct ExplorerDB {
-    multiverse: Multiverse<Ref>,
-    // This is kind of the same thing the multiverse holds (with Ref instead of BlockId)
-    // FIXME: The constructor of `ChainLength` is private, so querying this thing could be
-    // a problem
-    chain_length_to_hash: Lock<HashMap<ChainLength, Vec<HeaderHash>>>,
-    transaction_to_block: Lock<HashMap<FragmentId, HeaderHash>>,
-}
-
 impl ExplorerDB {
-    pub fn new() -> Self {
-        Self {
-            multiverse: Multiverse::<Ref>::new(),
-            chain_length_to_hash: Lock::new(HashMap::new()),
-            transaction_to_block: Lock::new(HashMap::new()),
-        }
-    }
+    pub fn bootstrap(block0: Block) -> Result<Self> {
+        // TODO: Here we should load from Storage to the current Head
 
-    pub fn store_ref(
-        &mut self,
-        new_block_ref: Ref,
-    ) -> impl Future<Item = GCRoot, Error = Infallible> {
-        let chain_length = new_block_ref.chain_length();
-        let header_hash = new_block_ref.hash();
+        let blocks = apply_block_to_blocks(Blocks::new(), &block0)?;
+        let epochs = apply_block_to_epochs(Epochs::new(), &block0)?;
+        let chain_lengths = apply_block_to_chain_lengths(ChainLengths::new(), &block0)?;
 
-        // Clone things to move into closures, this is just cloning locks
-        let mut map = self.chain_length_to_hash.clone();
-        let multiverse = self.multiverse.clone();
+        // XXX: I think the block0 doesn't have transactions
```

Comment:
it does, they are the initial funds (from ICO or whatever...)

Reply:
my mistake then. I'll change that